### DISCUSSION TITLE
Need help with a custom gpg-agent script

### DISCUSSION THREAD

**Discussion by 62832** [davmac314/dinit#194] [Q&A] [answered]
I'm currently in the process of writing a custom user service based on the one in the following [Artix Linux forum thread](https://forum.artixlinux.org/index.php/topic,3268), with which to start a supervised instance of `gpg-agent`. The way that I'm approaching this, however, is probably rather unorthodox.

For starters, I use GnuPG with a custom `GNUPGHOME` defined as opposed to the default `~/.gnupg` directory. This means that the GPG agent does not use the usual location for sockets, instead opting to use a hashed sub-directory of the standard `$XDG_RUNTIME_DIR/gnupg` directory. With this in mind, I'm trying to write my user service with a script that takes this into account and sets the socket path to listen on accordingly.

My current service definition looks like this:
```
type = process
load-options = sub-vars
logfile = $XDG_STATE_HOME/dinit/gpg-agent.log
socket-listen = $GPG_SOCKET_DIR/S.gpg-agent
command = $XDG_CONFIG_HOME/dinit.d/scripts/gpg-agent
```

While the script that this service runs is as follows:
```
#!/bin/sh
GPG_SOCKET_DIR="$XDG_RUNTIME_DIR/gnupg"

if [ -n "$GNUPGHOME ]; then
    GPG_SOCKET_DIR="${GPG_SOCKET_DIR}/$(gpgconf --list-dirs | grep "socketdir" - | sed 's:.*/::')"
fi

export GPG_SOCKET_DIR
dinitctl setenv GPG_SOCKET_DIR || :

mkdir -p "$GPG_SOCKET_DIR"
chmod 0700 "$GPG_SOCKET_DIR"

gpg-agent "${GNUPGHOME:+--homedir "$GNUPGHOME"}" "$@" --supervised
```

Of course, this service does not properly start up, instead hanging the terminal when attempting to start it. How might I be able to address this given what I currently have? I fear that I may be misunderstanding `dinitctl setenv` here among other things.

**Comment by davmac314** [comment, marked as answer]
One problem that stands out to me is that you are using `$GPG_SOCKET_DIR` in the service description file but it looks like `GPG_SOCKET_DIR` isn't set until the service runs (via `dinitctl setenv GPG_SOCKET_DIR` in the script). If the service description for a service uses a variable then the variable must be set at the time the service description is loaded.

I would have expected that this would cause opening the socket to fail and so the service would fail to start. Have you checked the service status, and the dinit log?

**Reply by 62832** [reply]
Indeed, my problem was that this wasn't set prior to starting the service. I've moved out the bulk of the code that sets the socket path to my `.bash_profile` to ensure it does get set prior and now the script works just fine.

Now I have the following which works. Under `.bash_profile`:
```
GNUPGSOCKET="$XDG_RUNTIME_DIR/gnupg/$(gpgconf --list-dirs | grep "socketdir" - | sed 's:.*/::')"
export GNUPGSOCKET
```

While the service itself is simply
```
type = process
load-options = sub-vars
logfile = $XDG_STATE_HOME/dinit/gpg-agent.log
socket-listen = $GNUPGSOCKET/S.gpg-agent
command = gpg-agent --homedir $GNUPGHOME --supervised
```
